Source/Core/Common/SPSCQueue.h

```diff
@@ -6,104 +6,181 @@
-  template <typename Arg>
-  void Push(Arg&& t)
+  void DecSize()
   {
-    // create the element, add it to the queue
-    m_write_ptr->current = std::forward<Arg>(t);
-    // set the next pointer to a new element ptr
-    // then advance the write pointer
-    ElementPtr* new_ptr = new ElementPtr();
-    m_write_ptr->next.store(new_ptr, std::memory_order_release);
+    AtomicSize::DecSize();
+    m_size.notify_one();
+  }
+};
+#endif
+};  // namespace detail
+
+template <typename T, typename SizeBase = detail::AtomicSize>
+class SPSCQueue final : public SizeBase
+{
+public:
+  SPSCQueue() = default;
+  ~SPSCQueue()
+  {
+    Clear();
+    delete m_read_ptr;
+  }
+
+  SPSCQueue(const SPSCQueue&) = delete;
+  SPSCQueue& operator=(const SPSCQueue&) = delete;
+
+  bool Empty() const { return SizeBase::Size() == 0; }
+
+  // The following are only safe from the "producer thread":
+  void Push(const T& arg) { Emplace(arg); }
+  void Push(T&& arg) { Emplace(std::move(arg)); }
+  template <typename... Args>
+  void Emplace(Args&&... args)
+  {
+    std::construct_at(&m_write_ptr->value.data, std::forward<Args>(args)...);
+
+    Node* const new_ptr = new Node;
+    m_write_ptr->next = new_ptr;
     m_write_ptr = new_ptr;
-    if (NeedSize)
-      m_size++;
+
+    SizeBase::IncSize();
   }
 
+  // The following are only safe from the "consumer thread":
+  T& Front() { return m_read_ptr->value.data; }
+  const T& Front() const { return m_read_ptr->value.data; }
+
   void Pop()
   {
-    if (NeedSize)
-      m_size--;
-    ElementPtr* tmpptr = m_read_ptr;
-    // advance the read pointer
-    m_read_ptr = tmpptr->next.load();
-    // set the next element to nullptr to stop the recursive deletion
-    tmpptr->next.store(nullptr);
-    delete tmpptr;  // this also deletes the element
+    assert(!Empty());
+
+    std::destroy_at(&Front());
+
+    Node* const old_node = m_read_ptr;
+    m_read_ptr = old_node->next;
+    delete old_node;
+
+    SizeBase::DecSize();
   }
 
-  bool Pop(T& t)
+  bool Pop(T& result)
   {
     if (Empty())
       return false;
 
-    if (NeedSize)
-      m_size--;
-
-    ElementPtr* tmpptr = m_read_ptr;
-    m_read_ptr = tmpptr->next.load(std::memory_order_acquire);
-    t = std::move(tmpptr->current);
-    tmpptr->next.store(nullptr);
-    delete tmpptr;
+    result = std::move(Front());
+    Pop();
     return true;
   }
 
-  // not thread-safe
   void Clear()
   {
-    m_size.store(0);
-    delete m_read_ptr;
-    m_write_ptr = m_read_ptr = new ElementPtr();
+    while (!Empty())
+      Pop();
   }
 
 private:
-  // stores a pointer to element
-  // and a pointer to the next ElementPtr
-  class ElementPtr
+  struct Node
   {
-  public:
-    ElementPtr() : next(nullptr) {}
-    ~ElementPtr()
+    // union allows value construction to be deferred until Push.
+    union Value
     {
-      ElementPtr* next_ptr = next.load();
-
-      if (next_ptr)
-        delete next_ptr;
-    }
+      T data;
```

Comment:
This is a cute trick, I didn't realize you could leave an union completely uninitialized. As far as I can tell [this is legal](https://en.cppreference.com/w/cpp/language/union#Member_lifetime), so yeah, cool.